src/features/terminal/utils.ts

```diff
@@ -1,19 +1,115 @@
 import * as path from 'path';
-import { Terminal, TerminalOptions, Uri } from 'vscode';
+import { Disposable, Terminal, TerminalOptions, Uri } from 'vscode';
 import { PythonEnvironment, PythonProject, PythonProjectEnvironmentApi, PythonProjectGetterApi } from '../../api';
-import { sleep } from '../../common/utils/asyncUtils';
+import { timeout } from '../../common/utils/asyncUtils';
+import { createSimpleDebounce } from '../../common/utils/debounce';
+import { onDidChangeTerminalShellIntegration, onDidWriteTerminalData } from '../../common/window.apis';
 import { getConfiguration, getWorkspaceFolders } from '../../common/workspace.apis';
 
 export const SHELL_INTEGRATION_TIMEOUT = 500; // 0.5 seconds
 export const SHELL_INTEGRATION_POLL_INTERVAL = 20; // 0.02 seconds
 
+/**
+ * Three conditions in a Promise.race:
+ * 1. Timeout based on VS Code's terminal.integrated.shellIntegration.timeout setting
+ * 2. Shell integration becoming available (window.onDidChangeTerminalShellIntegration event)
+ * 3. Detection of common prompt patterns in terminal output
+ */
 export async function waitForShellIntegration(terminal: Terminal): Promise<boolean> {
-    let timeout = 0;
-    while (!terminal.shellIntegration && timeout < SHELL_INTEGRATION_TIMEOUT) {
-        await sleep(SHELL_INTEGRATION_POLL_INTERVAL);
-        timeout += SHELL_INTEGRATION_POLL_INTERVAL;
+    if (terminal.shellIntegration) {
+        return true;
+    }
+
+    const config = getConfiguration('terminal.integrated');
+    const timeoutValue = config.get<number | undefined>('shellIntegration.timeout');
+    const timeoutMs = timeoutValue === undefined || -1 ? 5000 : timeoutValue;
+
+    const disposables: Disposable[] = [];
+
+    try {
+        const result = await Promise.race([
+            // Condition 1: Shell integration timeout setting
+            timeout(timeoutMs).then(() => false),
+
+            // Condition 2: Shell integration becomes available
+            new Promise<boolean>((resolve) => {
+                disposables.push(
+                    onDidChangeTerminalShellIntegration((e) => {
+                        if (e.terminal === terminal) {
+                            resolve(true);
+                        }
+                    }),
+                );
+            }),
+
+            // Condition 3: Detect prompt patterns in terminal output
+            new Promise<boolean>((resolve) => {
+                let dataSoFar = '';
+                const debounced = createSimpleDebounce(50, () => {
+                    if (dataSoFar && detectsCommonPromptPattern(dataSoFar)) {
+                        resolve(false);
+                    }
+                });
+                disposables.push(
+                    onDidWriteTerminalData((e) => {
+                        if (e.terminal === terminal) {
+                            dataSoFar += e.data;
+                            debounced.trigger();
+                        }
+                    }),
+                );
```

Comment:
nit: This would faster/more memory friendly if it was a string builder:


```suggestion
                const dataEvents: string = [];
                const debounced = createSimpleDebounce(50, () => {
                    if (dataEvents && detectsCommonPromptPattern(dataEvents.join(''))) {
                        resolve(false);
                    }
                });
                disposables.push(
                    onDidWriteTerminalData((e) => {
                        if (e.terminal === terminal) {
                            dataEvents.push(e.data);
                            debounced.trigger();
                        }
                    }),
                );
```